src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.cs

```diff
@@ -1246,107 +1252,248 @@ protected override Expression VisitMethodCall(MethodCallExpression methodCallExp
+                                _arrayCopyMethodInfo,
+                                previousKeyValuesVariable!,
+                                currentKeyValuesVariable!,
+                                Expression.Constant(currentKeyValuesCount - 1));
+
+                            var missingKeyValueAssignment = Expression.Assign(
+                                Expression.MakeIndex(
+                                    currentKeyValuesVariable,
+                                    _objectArrayIndexerPropertyInfo,
+                                    new[] { Expression.Constant(currentKeyValuesCount - 1) }),
+                                Expression.Convert(
+                                    Expression.Add(elementAccessExpression, Expression.Constant(1)),
+                                    typeof(object)));
+
+                            _variables.Add(currentKeyValuesVariable);
+                            _expressions.Add(currentKeyValuesArrayInitAssignment);
+                            _expressions.Add(keyValuesArrayCopyFromPrevious);
+                            _expressions.Add(missingKeyValueAssignment);
+                        }
+                    }
+
+                    var jsonElementValueExpression = Expression.Condition(
+                        Expression.MakeMemberAccess(
+                            currentJsonElementVariable!,
+                            _nullableJsonElementHasValuePropertyInfo),
+                        Expression.Convert(
+                            jsonElementAccessExpressionFragment,
+                            currentJsonElementVariable!.Type),
+                        Expression.Default(currentJsonElementVariable!.Type));
+
+                    currentJsonElementVariable = Expression.Variable(
+                        typeof(JsonElement?));
 
                     var jsonElementAssignment = Expression.Assign(
-                        jsonElementVariable,
+                        currentJsonElementVariable,
                         jsonElementValueExpression);
 
-                    _variables.Add(jsonElementVariable);
+                    _variables.Add(currentJsonElementVariable);
                     _expressions.Add(jsonElementAssignment);
-                    _existingJsonElementMap[(jsonColumnProjectionIndex, additionalPath[..index])] = jsonElementVariable;
-
-                    currentJsonElementVariable = jsonElementVariable;
-                }
-                else
-                {
-                    currentJsonElementVariable = exisitingJsonElementVariable2;
                 }
 
                 index++;
             }
-            while (index <= additionalPath.Length);
+            while (index <= jsonProjectionInfo.AdditionalPath.Length);
+
+            return (currentJsonElementVariable!, currentKeyValuesVariable!);
+
+            ParameterExpression ExtractAndCacheNonConstantJsonCollectionElementAccessValue(int index)
```

Comment:
I think the standard JSON naming is array (rather than the more general "collection", which doesn't imply ordering). If so we should align to that wherever we're referring to a JSON array (variable/method names, etc.)

```suggestion
            ParameterExpression ExtractAndCacheNonConstantJsonArrayElementAccessValue(int index)
```